lib/rubocop/ast/node_pattern/node.rb

```diff
@@ -0,0 +1,195 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    class NodePattern
+      # Base class for AST Nodes of a `NodePattern`
+      class Node < ::Parser::AST::Node
+        extend Forwardable
+        include ::RuboCop::AST::Descendence
+
+        ###
+        # To be overriden by subclasses
+        ###
+
+        def rest?
+          false
+        end
+
+        def capture?
+          false
+        end
+
+        # @return [Integer, Range] An Integer for fixed length terms, otherwise a Range.
+        # Note: `arity.end` may be `Float::INFINITY`
+        def arity
+          1
+        end
+
+        # Return [Symbol] either :ok, :raise, or :insert_wildcard
+        def in_sequence_head
+          :ok
+        end
+
+        ###
+        # Utilities
+        ###
+
+        # @return [Array<Node>]
+        def children_nodes
+          children.grep(Node)
+        end
+
+        # @return [Node] most nodes have only one child
+        def child
+          children[0]
+        end
+
+        # @return [Integer] nb of captures of that node and its descendants
+        def nb_captures
+          children_nodes.sum(&:nb_captures)
+        end
+
+        # @return [Boolean] returns true iff matches variable number of elements
+        def variadic?
+          arity.is_a?(Range)
+        end
+
+        # @return [Range] arity as a Range
+        def arity_range
+          a = arity
+          a.is_a?(Range) ? a : INT_TO_RANGE[a]
+        end
+
+        INT_TO_RANGE = Hash.new { |h, k| h[k] = k..k }
+        private_constant :INT_TO_RANGE
+
+        ###
+        # Subclasses for specific node types
+        ###
+
+        # Registry
+        MAP = Hash.new(Node)
+        def self.setup_registry
+          %i[sequence repetition rest capture predicate any_order function_call].each do |type|
+            MAP[type] = const_get(type.to_s.gsub(/(_|^)(.)/) { Regexp.last_match(2).upcase })
+          end
+          MAP.freeze
+        end
```

Comment:
Is this for lazy initialization?
Otherwise, is it the same as:
```ruby
MAP = {
  sequence: Sequence,
  repetition: Repetition,
  rest: Rest,
  capture: Capture,
  predicate: Predicate,
  any_order: AnyOrder ,
  function_call: FunctionCall
}.freeze
```
For me, as a human, a literal map is easier on the eye.

Reply:
Agreed, probably best to simply list them all this way 👍